packages/loopring_v3/contracts/impl/agents/FastWithdrawalAgent.sol

```diff
@@ -0,0 +1,203 @@
+/*
+
+  Copyright 2017 Loopring Project Ltd (Loopring Foundation).
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+pragma solidity ^0.5.11;
+pragma experimental ABIEncoderV2;
+
+import "../../iface/IExchangeV3.sol";
+
+import "../../lib/Claimable.sol";
+import "../../lib/ERC20SafeTransfer.sol";
+import "../../lib/MathUint.sol";
+import "../../lib/ReentrancyGuard.sol";
+import "../../lib/SignatureUtil.sol";
+import "../../lib/EIP712.sol";
+
+
+/// @title Fast withdrawal agent implementation.
+///
+///        A mediator/liquidity provider (whoever provides the funds immediately) can call
+///        `executeFastWithdrawals` to provide users immediately with funds onchain.
+///        This allows the security of the funds to be handled by any EOA or smart contract.
+///
+///        Users that want to make use of this functionality have to
+///        authorize this contract as their agent.
+///
+///        TODO: generalize this (here or another agent) with any function call
+///        a user wants to do to be able to simulate completely onchain smart wallet with
+///        funds stored in the Loopring protocol.
+///
+/// @author Brecht Devos - <brecht@loopring.org>
+contract FastWithdrawalAgent is ReentrancyGuard
+{
+    using AddressUtil       for address;
+    using AddressUtil       for address payable;
+    using ERC20SafeTransfer for address;
+    using MathUint          for uint;
+    using SignatureUtil     for bytes32;
+
+    struct FastWithdrawal
+    {
+        address exchange;
+        address from;
+        address to;
+        address token;
+        uint    amount;
+        address feeToken;
+        uint    fee;
+        uint    validUntil;
+        bool    onchainFeePayment;
+        bytes   signature;
+    }
+
+    // Replay protection
+    mapping (bytes32 => bool) txHash;
+
+    // EIP712
+    bytes32 constant public FASTWITHDRAWAL_TYPEHASH = keccak256(
+        "FastWithdrawal(address exchange,address from,address to,address token,uint256 amount,address feeToken,uint256 fee,bool onchainFeePayment,uint256 validUntil)"
+    );
+    bytes32 public DOMAIN_SEPARATOR;
+
+    constructor()
+        public
+    {
+        //DOMAIN_SEPARATOR = EIP712.hash(EIP712.Domain("FastWithdrawalAgent", "1.0", address(this)));
+    }
+
+    function executeFastWithdrawals(FastWithdrawal[] memory fastWithdrawals)
```

Comment:
Maybe add a comment like this:
This method is expected to be called by a mediator to facilitate fast withdrawals.

Questions:


1. All FastWithdrawals are offchain requests, right?

1. how can a mediator get the `fastWithdrawals` data, from the operator? If so, the mediator may be strongly associated with the operator and has to trust the operator. If the operator lies, then the mediator may have sent the fund but end up with not being paid on layer 2 (for example, the user doesn't have enough layer2 balance at all, but the operator told the mediator the user has).

There may be some cornor cases. For example: if the mediator didn't respond quickly enough, the operator may think the mediator doesn't want to fulfil the fastwithdrawal requests, then he will free layer2 balance locked by the fast withdrawal requests, if later this fund got used by trades, and the mediator now pays the users with layer1 balance, he will end up with a loss. 

The current design handles this by using the validUnitl timestamp, I guess. Right?